### DISCUSSION TITLE
Proposed UC Roadmap CY2024Q4

### DISCUSSION THREAD

**Discussion by dennyglee** [unitycatalog/unitycatalog#411] [Ideas]
> [!IMPORTANT]
> This roadmap discussion is outdated and the proposed roadmap no longer valid.
> Please find the current roadmap in the [`roadmap.md`](https://github.com/unitycatalog/unitycatalog/blob/main/roadmap.md) file in the repository root.

There have been a lot of great questions about the UC roadmap and based on the various discussions, here is the *proposed* CY2024Q4 roadmap.  Currently, we are *targeting* a quarterly release cadence starting with the September 0.2 release.  As always, features may move in/out of milestones pending available resources and priorities.

## Core
Core Unity Catalog functionality

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| Catalog                          | ![done]     | ![done]     | ![done]     | ![done]     | ![done]     |
| Schema                           | ![done]     | ![done]     | ![done]     | ![done]     | ![done]     |
| Managed Catalog                  |             |             | ![done]     | ![done]     | ![done]     | 
| Managed Schema                   |             |             | ![done]     | ![done]     | ![done]     | 
| Apache Spark™                    |             |             |             | ![open]     | ![done]     | 
| Storage Credentials              |             |             |             | ![done]     | ![done]     | 
| External Location                |             |             |             | ![done]     | ![done]     | 
| Multi-tenancy                    |             |             |             | ![done]     | ![done]     | 
| Support for on-prem storage      |             |             |             |             | ![done]     | 

## Governance
Unity Catalog Governance features

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| OAuth                            |             | ![done]     | ![done]     | ![done]     | ![done]     |
| Local identity management        |             | ![done]     | ![done]     | ![done]     | ![done]     |
| Temporary credential vending for tables      |             | ![done]     | ![done]     | ![done]     | ![done]     |
| Temporary credential vending for volumes     |             | ![done]     | ![done]     | ![done]     | ![done]     |
| Temporary credential vending for models      |             | ![done]     | ![done]     | ![done]     | ![done]     |
| Basic grants                     |             | ![done]     | ![done]     | ![done]     | ![done]     |
| SSO with external IdP            |             |             |             |             | ![open]     |
| Machine identities (SPs) and authentication  |             |             |             |             | ![open]     |
| Identity sync from another IdP with SCIM     |             |             |             |             | ![open]     |
| Permission parity with managed UC            |             |             |             |             | ![open]     |
| RBAC                             |             |             |             |             | ![open]     |
| ABAC                             |             |             |             |             | ![open]     |
| Row and column access policies   |             |             |             |             | ![open]     |
| Lineage                          |             |             |             |             | ![open]     |
| Auditing                         |             |             |             |             | ![open]     |

## Server production-readiness
Unity Catalog production readiness

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| Monitoring and Telemetry         |             |             |             |             | ![open]     |
| Database schema upgrades         |             |             |             |             | ![open]     |
| Change events                    |             |             |             |             | ![open]     |

## Tables
Unity Catalog Table features

<table>
<tr>
    <td><b>Feature</b></td>
    <td><b>Area</b></td>
    <td><b>v0.1</b></td>
    <td><b>v0.2</b></td>
    <td><b>v0.3</b></td>
    <td><b>v0.4</b></td>
    <td><b>v0.5+</b></td>
<tr>
    <td rowspan=3>External table reads & writes</td>
    <td>API + Server</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Spark Integration</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Delta Integration</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td rowspan=2>Managed Delta table reads</td>
    <td>API + Server</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Delta + Spark Integration</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td rowspan=3>Managed Delta table creates/writes with coordinated commits</td>
    <td>API + Server</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Delta-Spark Integration</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Delta Kernel Integration</td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td rowspan=2>Delta Uniform tables with read as Iceberg via Iceberg REST API</td>
    <td>API + Server</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChangesGrey.svg" alt="semi-done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChangesGrey.svg" alt="semi-done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Delta Integration</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td colspan=2>Iceberg tables with CRUD operations</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td colspan=2>Multi-engine data types for column definitions</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
</table>

## Views
Unity Catalog View features

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| Basic Spark SQL flavor views     |             |             |             | ![open]     | ![done]     |
| Multi-dialect views              |             |             |             |             | ![done]     |
| Iceberg view support             |             |             |             |             | ![done]     |
| Materialized views               |             |             |             |             | ![done]     |
| Streaming tables                 |             |             |             |             | ![done]     |
| Shallow clones                   |             |             |             |             | ![done]     |

## Non-tabular and AI assets
Unity Catalog Non-tabular and AI assets

<table>
<tr>
    <td><b>Feature</b></td>
    <td><b>Area</b></td>
    <td><b>v0.1</b></td>
    <td><b>v0.2</b></td>
    <td><b>v0.3</b></td>
    <td><b>v0.4</b></td>
    <td><b>v0.5+</b></td>
<tr>
    <td rowspan=3>Functions (SQL UDFs, Python UDFs)</td>
    <td>API + Server</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>ML integrations with advanced Python SDK</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Spark Integration</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td colspan=2>Multi-engine functions (SQL)</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
</tr>
<tr>
    <td colspan=2>Remote functions</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
</tr>
<tr>
    <td rowspan=2>External volumes</td>
    <td>API + Server</td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Spark Integration</td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td rowspan=2>Managed Volumes</td>
    <td>API + Server</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>Spark Integration</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td rowspan=2>Models and model versions</td>
    <td>API + Server</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td>MLflow Integration</td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg" alt="done"/></td>
</tr>
<tr>
    <td colspan=2>Feature tables</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
</tr>
<tr>
    <td colspan=2>Data monitors</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td><img src="https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg" alt="open"/></td>    
</tr>
</table>

## Sharing
Unity Catalog Sharing features

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| Delta Sharing integration        |             |             |             |             | ![open]     |
| Shares                           |             |             |             |             | ![open]     |
| Recipients                       |             |             |             |             | ![open]     |
| Providers                        |             |             |             |             | ![open]     |

## Federation
Unity Catalog Federation features

| Feature                          | v0.1        | v0.2        | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :---------: | :---------: | :---------: | :---------: | :---------: |
| Connections                      |             |             |             |             | ![open]     |
| Foreign objects (catalogs, schemas, tables)                    |             |             |             |             | ![open]     |
| Support for different data sources: JDBC, Iceberg REST, HMS    |             |             |             |             | ![open]     |

## UI
Unity Catalog UI features

| Feature                          | v0.1         | v0.2         | v0.3        | v0.4        | v0.5+       | 
| -------------------------------- | :----------: | :----------: | :---------: | :---------: | :---------: |
| Basic UI                         | ![semi-done] | ![done]      | ![done]     | ![done]     | ![done]     |
| Delete volumes, function, schema, catalog    | ![semi-done] | ![done]     | ![done]     | ![done]     | ![done]     |
| Update volumes, function, schema, catalog    | ![semi-done] | ![done]     | ![done]     | ![done]     | ![done]     |
| Login and Authentication         |              | ![semi-done] | ![done]     | ![done]     | ![done]     |
| OpenAPI spec to generate types   |              | ![semi-done] | ![done]     | ![done]     | ![done]     |
| Publish via npm                  |              | ![semi-done] | ![done]     | ![done]     | ![done]     |

**Note**: We are still working on proposals for UI features after Milestone v0.2

[open]: https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/IssueNeutral.svg
[semi-done]: https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChangesGrey.svg
[done]: https://cdn.jsdelivr.net/gh/Readme-Workflows/Readme-Icons@main/icons/octicons/ApprovedChanges.svg

**Comment by ni-mi** [comment]
Hi @dennyglee - looks very interesting!

Two questions:
1.Regarding timelines... The title is Proposed UC Roadmap CY2024Q4 - while I see you mention milestones up to v0.5.
Are all those versions expected until EOY? I presume not and v0.5 will be around q2-3 2025... What is the proposed timeline?

2.Regarding compatibility of Databricks UC with OSS UC: there is a new public preview for credentials vending within Databricks - but it does not include managed tables yet while here tables credentials are checked for v0.2, managed tables writing and creation is checked for v0.3 - while I couldn't find timeline for Databricks UC managed tables creation and writing support from external Spark. What are the timelines for those?

Thanks!
Nimrod

**Reply by dennyglee** [reply]
_Per timelines_
Oh, the idea is that we re-assess the roadmap every quarter so we can communicate what we are prioritizing for this and subsequent quarters.   You are correct that _currently_ (i.e., as of today October 7th, 2024), the proposed v0.5 would be 2025Q2/Q3.  But each quarter we will want to re-assess if this makes sense and apply/prioritize accordingly. 

_Databricks UC_
Oh, best to track that with the Databricks roadmap as while we try to stay informed of this, as UC is an OSS project, we will prioritize based on community feedback such as this forum.

**Comment by dwolfson** [comment]
Hi @dennyglee  - a couple of thoughts..

1) Most of the items under your governance category seem more like security? Perhaps there should be both a security category and a governance category?

2) Should there be a category for documentation and samples, in particular, I could imagine sample Egeria/Marquez/Airflow/DBT etc integrations?

**Reply by dennyglee** [reply]
Thanks for your patience @dwolfson 

1. That's fair - I think it was just because it's still early I grouped them together.  I think as we progress, we can start breaking this out into a separate categories, eh?!
2. Sure, I think we can do that for the next rev of the roadmap as well.  Or do you think even with the current roadmap of 0.3 it would make sense for us to have that category?

**Reply by dwolfson** [reply]
Thanks @dennyglee, on the second point...  yes, I think it does make sense as there are already quite a few integrations mentioned in the documentation such as DuckDB and PuppyGraph. Having a category for integrations with documentation and samples would provide a place for the community to contribute artifacts demonstrating and documenting different integrations...

**Comment by cornelcreanga** [comment]
Hello @dennyglee - I see that "Iceberg tables with CRUD operations" is scheduled in for v0.4, so I imagine this mean end of Q1. Are there any plans to prioritize that in order to add it to v0.3? We moved a large datalake from Hive to Iceberg and now we would like to replace the Hive Metastore with something else - personally I would love to choose Unity vs Polaris/Nessie but I don't know if we can wait so long. Thanks a lot.

**Reply by dennyglee** [reply]
Thanks @cornelcreanga - I definitely hear you (and others) loud and clear!  I think the concern is the number of available resources to get this done.  I take it that using UniForm with Iceberg is not a good short-term solution here?  cc: @michelleon @tdas

**Reply by michelleon** [reply]
Hey @cornelcreanga implementing CRUD operations for Iceberg tables is high priority for us in UC. It's the top of the list to get pulled into 0.3, but as Denny said it's a matter of available folks and prioritization. Working on it!

**Reply by cornelcreanga** [reply]
@dennyglee - I don't think so, it would be impractical to work with multiple catalogs, one for write/one for reads. Migrating to delta is not an option neither (we have ~ 1 quadrillion events for now). However I guess we'll be able to start with another REST catalog/switch to Unity at some moment.
@michelleon  - thanks. Reading on https://docs.databricks.com/en/delta/uniform.html#limitations I see that positional deletes are also not working, is this up to date ? Most of our updatable tables cases are using mor instead of copy on write..are they not going to be readable?

**Comment by apd-bbrewington** [comment]
I might have missed the memo, but what does the "![](https://camo.githubusercontent.com/6392cdeaf83d2849b20c5423367196fd9d4604ad57213cb84904c2b67c9caba1/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f526561646d652d576f726b666c6f77732f526561646d652d49636f6e73406d61696e2f69636f6e732f6f637469636f6e732f49737375654e65757472616c2e737667)" icon mean?

**Reply by dennyglee** [reply]
Oh that icon represents we will try to do this for that particular release but we currently do not have enough folks to tackle it.  But we're trying really hard with the current set of contributors to try to get it out for that release cycle.

**Comment by DmitriyAlergant** [comment]
@dennyglee can you explain the development process of UC OSS.  Are all of these features basically independently redeveloped from scratch, or is the existing Databricks managed UC code being back-ported and released to OSS?     From this roadmap it looks like at least a year [optimistically] to even approach feature-parity with the (current capability level) Databricks UC...

When Matei announced UC open-sourcing on DAIS, it kind of sounded as open-sourcing the existing code-base, rather then starting a parallel project of reimplementing from scratch?

**Reply by benbauer-draeger** [reply]
Hey, this is the most upvoted question and there is just no reply. Feels like open sourcing UC was just PR :(
What about adding the newly related features announced at the AI Summit 25? https://www.databricks.com/blog/whats-new-databricks-unity-catalog-data-ai-summit-2025

**Reply by kshitij-g** [reply]
not sure, why the UC OSS is focused on first converting it to a MCP server with a registry of AI functions, tools and other stuff but not bringing the UC OSS features at par with databricks UC.

**Comment by ni-mi** [comment]
@dennyglee 
I see that external locations and Storage Credentials are planned for 0.4.0 - while volumes are already supported...
How is that?
Volumes are based on external locations - and use storage credentials to work...

Thanks!
Nimrod

**Comment by ni-mi** [comment]
Hi @dennyglee @michelleon 
I see that 0.3 release is scheduled to have Managed Delta table creates/writes with coordinated commits

coordinated commits is not yet finalized in Delta, and at least from the [Preview version](https://delta.io/blog/delta-lake-4-0/) it seems that it will only work on Spark 4.0 - which is also not finalized...

Does that mean that UC 0.3 will not be released until Spark 4.0 - and will support only Spark 4.0 and up?

Thanks!
Nimrod

**Reply by dennyglee** [reply]
Great call out @ni-mi , I think that's up for discussion. Right now, when reviewing most of the code base, a lot of the changes are based on the `unitycatalog-ai` which has (for now) it's own versioning scheme.  Let's plan on discussing in early January what we should include in 0.3 release.  We shouldn't have a light release for the sake of versioning but at the same time, I would like to see us having a more predictive release cadence.

**Reply by ni-mi** [reply]
@dennyglee we are in early January - is there any update regarding writing to managed tables?

**Comment by nimrod-doubleverify** [comment]
@dennyglee What about updating the versions timelines - and what will be in each version?

**Comment by talebzeghmi** [comment]
Does the v3 release https://github.com/unitycatalog/unitycatalog/releases/tag/ai-v0.3.0 mean that the v0.3 features above, namely that Managed Delta table creates/writes and Volumes are supported?

**Reply by dennyglee** [reply]
Not exactly the tag is `ai-v0.3.0` which is part of the `unitycatalog-ai` release cadence.  We've been doing a lot of work on the AI side while we get the PRDs ready for review around managed tables and coordinated commits.

**Reply by ni-mi** [reply]
@dennyglee I think that the question is what is the proposed timeline for managed tables and volume writes...

**Reply by steffen030** [reply]
@dennyglee It looks like the outlines roadmap doesn't seem to be relevant anymore. It would be still interesting to understand if some of the basic features outlines in the roadmap above are still on the roadmap and for which release they are targeted.

**Reply by dennyglee** [reply]
@steffen030 Completely fair.  I think we're all a bit underwater for the upcoming DAIS25 conference where we should have a lot more clarity afterward.  Apologies for the delay, a lot has been going on the last few months, eh?!

**Reply by kshitij-g** [reply]
now that DAIS25 is over, is there any update? very basic features are missing in unity catalog oss vs databricks unity catalog. When can we expect parity at least between the two, let alone the additional features that you guys plan to develop, included upgrading to spark / delta 4.0.0

**Reply by brijesh1998** [reply]
Any update on the updated roadmap? It’s been a month (almost) since DAIS 2025, and many of us were expecting some clarity on how announcements would reflect in the OSS roadmap.

I truly appreciate the work that’s gone into the current direction, especially around `unitycatalog-ai`. That said, I also agree with @kshitij-g — since the OSS version of UC is being built from scratch, it might be more impactful for community if we first focus on getting core UC features in place.
@dennyglee — I am hoping to hear back from you on this!
